include/boost/json/detail/value_to.hpp

```diff
@@ -67,152 +69,177 @@ void
 // array
 inline
 array
-tag_invoke(
+value_to_impl(
     value_to_tag<array>,
-    value const& jv)
+    value const& jv,
+    array_conversion_tag)
 {
     return jv.as_array();
 }
 
 // string
 inline
 string
-tag_invoke(
+value_to_impl(
     value_to_tag<string>,
-    value const& jv)
+    value const& jv,
+    string_conversion_tag)
 {
     return jv.as_string();
 }
 
 // bool
 inline
 bool
-tag_invoke(
+value_to_impl(
     value_to_tag<bool>,
-    value const& jv)
+    value const& jv,
+    bool_conversion_tag)
 {
     return jv.as_bool();
 }
 
 // integral and floating point
-template<class T, typename std::enable_if<
-    std::is_arithmetic<T>::value>::type* = nullptr>
+template<class T>
 T
-tag_invoke(
+value_to_impl(
     value_to_tag<T>,
-    value const& jv)
+    value const& jv,
+    number_conversion_tag)
 {
     return jv.to_number<T>();
 }
 
-//----------------------------------------------------------
-// Use generic conversion
+// null-like conversion
+template<class T>
+T
+value_to_impl(
+    value_to_tag<T>,
+    value const& jv,
+    null_like_conversion_tag)
+{
+    if( jv.is_null() )
+        return T();
+    detail::throw_invalid_argument(
+        "source value is not null", BOOST_JSON_SOURCE_POS);
+}
 
 // string-like types
-// NOTE: original check for size used is_convertible but
-// MSVC-140 selects wrong specialisation if used
-template<class T, typename std::enable_if<
-    std::is_constructible<T, const char*, std::size_t>::value &&
-    std::is_convertible<decltype(std::declval<T&>().data()), const char*>::value &&
-    std::is_integral<decltype(std::declval<T&>().size())>::value
->::type* = nullptr>
+template<class T>
 T
-value_to_generic(
-    const value& jv,
-    priority_tag<3>)
+value_to_impl(
+    value_to_tag<T>,
+    value const& jv,
+    string_like_conversion_tag)
 {
     auto& str = jv.as_string();
-    return T(str.data(), str.size());
+    return T(str.subview());
 }
 
-// map-like containers; should go before other containers in order to be able
-// to tell them apart
-template<class T, typename std::enable_if<
-    has_value_to<typename map_traits<T>::pair_value_type>::value &&
-        std::is_constructible<typename map_traits<T>::pair_key_type,
-    string_view>::value>::type* = nullptr>
+// map-like containers
+template<class T>
 T
-value_to_generic(
-    const value& jv,
-    priority_tag<2>)
+value_to_impl(
+    value_to_tag<T>,
+    value const& jv,
+    map_like_conversion_tag)
 {
-    using value_type = typename
-        container_traits<T>::value_type;
     const object& obj = jv.as_object();
     T result;
-    try_reserve(result, obj.size(), priority_tag<2>());
-    for (const auto& val : obj)
-        result.insert(value_type{typename map_traits<T>::
-            pair_key_type(val.key()), value_to<typename
-                map_traits<T>::pair_value_type>(val.value())});
+    detail::try_reserve(result, obj.size(), reserve_implementation<T>());
+    std::transform(obj.begin(), obj.end(),
```

Comment:
Does this mean we are including `<algorithm>` now?

Reply:
We were already including it